### ISSUE TITLE
Coverage & performance re-verification at the publish pin (publish #58, Track 3)

### ISSUE THREAD

**Issue by OriPekelman** [OriPekelman/toy#61]
Part of #58. What we claim, we have verified — at the pin we publish against.

- Re-run model-parity gates for the README table: SmolLM2, Llama 3, Qwen 2.5/3, Mistral, Gemma 2, OLMoE; record revs + dates in docs/coverage.md; name unsupported things explicitly (a "not yet" list is a feature).
- Backends matrix in docs (and one line in README): CPU full · CUDA train/infer/eval, serve CPU-only · Metal pending #27 (Mac pass is the gate for any Metal claim).
- Refresh docs/bench-vs-pytorch.md at the pin; include the honest caveats (whole-graph speedups vs toy-shape CUDA losses; cf. the never-extrapolate lesson).
- Primitives: docs/coverage.md op table re-checked against the pin.

**Comment by OriPekelman** [comment]
## Re-verification pass — 2026-06-11, gx10

**Pin:** toy `d605aea` / spinel `a699cf9` (`/tmp/spinel-a699cf9-serve`) / ggml `41e7949`; serve leg at spinel `f6d5eef` (spinel-dev#14). PR: #74 (branch `docs/toy61-coverage-reverify`, pushed to both remotes).

**Method per row:** greedy `toy infer` (n=8), CUDA asserted token-identical to CPU; Q8 asserted token-identical to F32 where both exist; GPT-2 family via the tinynn logits-parity probes against the recorded HF/PyTorch reference (`data/ref_logits.txt`) and the May-recorded CUDA logits.

### Model matrix (row × precision × backend)

| Model | F32 CPU | Q8 CPU | CUDA F32 | CUDA Q8 | Metal |
|---|---|---|---|---|---|
| DistilGPT-2 | ✓ PT-parity (max&#124;Δ&#124;=5.4e-3, argmax+top5 match) | — | ✓ bit-identical to CPU **and** to May recording | — | ‡ |
| GPT-2 small | ✓ KV parity | — | ✓ KV CPU≡CUDA bit-identical ≡ May recording | — | ‡ |
| SmolLM2-135M | ✓ byte-exact infer+eval gate baselines | ✓ ids ≡ F32 | ✓ bit-identical (gate CUDA arm) | — | ✓ stands on 2026-05-31 Mac pass, NOT re-run (#27) |
| SmolLM2-360M | ✓ | — | ✓ ids ≡ CPU | — | ‡ |
| TinyLlama-1.1B | ✓ coherent (tok) | ✓ ids ≡ F32 | ✓ ids ≡ CPU | — | ‡ |
| Llama-3.2-1B | ✓ coherent (tok) | — | ✓ ids ≡ CPU | — | ‡ |
| Llama-3.2-3B | ✓ | — | ✓ ids ≡ CPU | — | ‡ |
| Mistral-7B-v0.2 | ✓ coherent (29 GB f32-tok) | ✓ | — (not claimed) | ✓ ids ≡ CPU | — |
| Qwen2.5-0.5B | ✓ coherent (tok) | ✓ | ✓ ids ≡ CPU | — | ‡ |
| Qwen2.5-1.5B | ✓ | ✓ ids ≡ F32 | ✓ | **† → ✓ UPGRADED** — d_ff%512 abort no longer reproduces at ggml 41e7949; ids ≡ CPU | ‡ |
| Qwen2.5-3B | ✓ | ✓ | ✓ | **† → ✓ UPGRADED** (same) | ‡ |
| Qwen2.5-7B | ✓ (30 GB, one run) | ✓ ids ≡ F32 | — (not claimed) | ✓ ids ≡ CPU | — |
| Qwen3-0.6B | ✓ coherent | — | **✗ FAILED** (see below) | — | ‡ |
| Qwen3-1.7B | ✓ coherent | **SKIPPED-no-model** (no Q8 GGUF on box; refill = >2GB HF download) | **✗ FAILED** | — | ‡ |
| Qwen3-4B | ✓ coherent | — | — (not claimed) | — | ‡ |
| OLMoE-1B-7B | — | ✓ coherent + `gate-moe-kquant` PASS (Q4_K_M, distinct=10/10) | — | — | ‡ |
| Gemma 2-2b-it | — | ✓ coherent (SPM-Unigram path) | — | — | ‡ |

### FAILED — Qwen3 on CUDA (both sizes), reported before downgrading
CPU coherent, CUDA degenerate, verbatim:
```
qwen3-0.6b cpu : Once upon a time, there was a man who lived in
qwen3-0.6b cuda: Once upon a time Answer Answer Answer Instructions Instructions Instructions Instructions Answer
qwen3-1.7b cpu : Once upon a time, the first time I saw the movie
qwen3-1.7b cuda: Once upon a time::::::::
```
`toy eval --device cuda` top-5 fully disjoint from CPU (0.6B: cpu `[220,68,0,8958,71]` vs cuda `[9,15846,12478,17980,15081]`). **Not an a699cf9 regression** — main-tree `libexec/toy-infer-cuda` built 2026-06-06 reproduces. Qwen3 is the only QK-norm + explicit-head_dim arch in the CUDA matrix (prime suspects: per-head qk-norm slicing / non-square w_o in the generated `llama_kv_engine_cuda.rb`). Table downgraded to ✗ with footnote. **Needs its own issue** (I couldn't file one from this run — permissions).

### Gates at a699cf9 — all PASS
infer (+CUDA arm, bit-identical) · eval (+CUDA arm, top-k identical) · train from-scratch/lora/warm-start/vit (byte-exact) · train-cuda ×3 (byte-exact) · gpt2-train CPU+CUDA (byte-exact) · full-finetune · ckpt-roundtrip · run-log · lmc · mixed-precision (f16 0.24% rel) · moe-kquant · poly-degrade (no new emit-0) · verify-mirrors · compute-surface CPU+CUDA. Serve at f6d5eef: `serve_gate` + `serve_events_gate` PASS byte-for-byte.

**coverage-check FAILED on main** (stale docs/coverage.md): 4 newly-bound ops (out_prod/cont_2d/permute/im2col/conv_2d), gelu autodiff, tanh-on-Metal, +33 infra wrappers missing; AND regenerating would have reverted the corrected MUL_MAT_ID prose (it was hand-edited into the generated file). Fixed in the generator + regenerated (PR commit 1).

### Bench drift (vs baselines recorded ~May 24)
`make bench` (CPU) — all ok, none regressed:
- lora_step b1 56.0 ms (-3.6%), b4 106.5 ms (+2.0%); steady-state b1 50.7 ms (-5.3%)
- infer 85.4 tok/s (**+32.6%** — faster), step 11.7 ms (-24.6%)
- tokenizer 1.41 us/tok (-4.9%)

`bench-vs-pytorch`: train ratio **0.995×** (budget 1.031, ok — parity holds). **Infer ratio NOT MEASURABLE**: `demos/qwen25_bench_cuda` segfaults —
```
Thread 1 received signal SIGSEGV in sp_gc_mark ()
#4 sp_PtrArray_new_scan #5 sp_SmolLM2KVCuda_cls_decode_step
```
Pre-existing (May 31 main-tree binary segfaults identically; budget was recorded May 24). Worse, `check_vs_pytorch.rb` returns `nil` for the missing metric, silently skips the ratio, and prints **ok** — fail-loud violation. Doc retires the 1.38× decode claim until fixed. **Needs its own issue** too.

### What the final-pin re-run needs
1. Re-run `prep/{infer,eval,train,train_cuda,gpt2_train*,serve*}_gate.rb` + `make coverage-check verify-mirrors bench` at the final pin and bump the three provenance stamps (models.md header, gen_coverage.rb provenance line + regenerate, bench-vs-pytorch.md) — everything else in the docs is already rev-keyed.
2. Re-check Qwen3-CUDA (✗ rows) — if the fix lands first, flip back with the new date.
3. Decide the `qwen25_bench_cuda` segv (spinel GC in `decode_step`) before re-recording the infer-vs-PT ratio.
4. Qwen3-1.7B Q8: either re-quantize on a box that has the HF weights or drop the carried-over ✓.
5. Metal rows stay Mac-gated (#27).

**Comment by OriPekelman** [comment]
Landed via PR #74 (main `e3452cf`). 15/17 models fully re-verified at toy d605aea / spinel a699cf9 / ggml 41e7949 with provenance stamped in docs; Qwen2.5-1.5B/3B CUDA Q8 **upgraded** (the † d_ff%512 footnote was stale at ggml 41e7949); Qwen3 CUDA F32 **downgraded** → #76; bench infer leg dead + silent-skip checker → #77. CPU infer +32.6% vs baseline; train ratio 0.995× vs PyTorch. The final-pin re-run folds into #62's gate sweep (cheap diff against these stamps). Metal rows remain Mac-gated (#27, refreshed brief posted).
